### PR TITLE
Edit/jetpack security plan recommendation

### PR DESCRIPTION
**Pull request by CodeyGuyDylan**
#### Summary

This changes the `security-plan` recommendation to a Jetpack Protect recommendation that auto-installs Protect when the CTA is selected

#### Changes proposed in this Pull Request:

This PR switches out the `security-plan` recommendation with a new `protect` one. Instead of linking to a blog post about security, this CTA will auto-install Jetpack Protect. The body of the recommendation has been rewritten to be related to Jetpack Protect and links to the [Jetpack Protect landing page](https://jetpack.com/protect)

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
P2: pbNhbs-3dM-p2

#### Does this pull request change what data or activity we track or use?

No additional tracking, these recommendations use the same set of tracking rules, this one is just going to be tracked as the 'protect' recommendation instead of the 'security-plan' one.

#### Testing instructions:

1. On your Jetpack test site, start by resetting the Jetpack options. If you are using a development version of Jetpack, this option should be available in the footer on the Jetpack Dashboard.
2. In the code, navigate to `projects/plugins/jetpack/_inc/client/recommendations/feature-utils.js` to line 118. Here, change the `jetpack-protect` plugin being installed to `protect` so we install the Protect plugin with the most recent changes that have not been released yet ( this PR will have to wait until Protect releases again in order to work correctly )
3. Navigate to the plugins library and install and active any plugin that you want.
4. Navigate back to the Jetpack dashboard, you should see a red "1" bade next to the "Recommendations" navigation item indicating that a new recommendation is available.
image
5. Click on the recommendations navigation item. You should now see the recommendation for Jetpack Protect. There should be a green "New" badge showing on this step.
image
6. Click on the `Learn More` external link and confirm it goes to https://jetpack.com/protect
7. Click on the `Get Jetpack Security` CTA and confirm it takes you to the checkout with Jetpack Security in the cart
Screen Shot 2022-08-22 at 9 29 35 AM
8. Finally, click on the main CTA. It should take you to the Summary and after a couple of sections, show Jetpack Protect installed with a CTA to go to the settings
image
9. Click on the Settings CTA and it should take you to the Jetpack Protect Dashboard
image

### CODE REVIEW

File: projects/plugins/jetpack/_inc/client/recommendations/prompts/prompt-layout/index.jsx
Comment (CodeyGuyDylan) [inline projects/plugins/jetpack/_inc/client/recommendations/prompts/prompt-layout/index.jsx:90]: This was a react error being thrown, decided to fix it while in this code

File: projects/plugins/jetpack/_inc/client/recommendations/sidebar/product-spotlight/index.jsx
Comment (CodeyGuyDylan) [inline projects/plugins/jetpack/_inc/client/recommendations/sidebar/product-spotlight/index.jsx]: Another react error fix

File: projects/plugins/jetpack/_inc/client/recommendations/sidebar/product-spotlight/index.jsx
Comment (CodeyGuyDylan) [inline projects/plugins/jetpack/_inc/client/recommendations/sidebar/product-spotlight/index.jsx:72]: Another react error fix

File: projects/plugins/jetpack/_inc/client/recommendations/constants.js
Comment (jboland88) [inline projects/plugins/jetpack/_inc/client/recommendations/constants.js]: Any reason not to change the constant name as well?

File: projects/plugins/jetpack/_inc/client/recommendations/constants.js
Comment (jboland88) [inline projects/plugins/jetpack/_inc/client/recommendations/constants.js:12]: I'm wondering if we will need to keep the old recommendation slug if only to redirect it to the new one. I think an error might get thrown if this recommendation is set as the current step in the state.

File: projects/plugins/jetpack/_inc/client/recommendations/sidebar/product-spotlight/index.jsx
Comment (jboland88) [reply projects/plugins/jetpack/_inc/client/recommendations/sidebar/product-spotlight/index.jsx]: Thanks for the fix here!

super nitpick - IIRC the 2nd variable in a map function is the index of the array item. IMO `id` implies another kind of unique identifier instead of the index. Could we change that variable name?

File: projects/plugins/jetpack/_inc/lib/class-jetpack-recommendations.php
Comment (jboland88) [inline projects/plugins/jetpack/_inc/lib/class-jetpack-recommendations.php]: Any reason not to change the constant name here as well?

File: projects/plugins/jetpack/_inc/client/recommendations/constants.js
Comment (jboland88) [reply projects/plugins/jetpack/_inc/client/recommendations/constants.js:12]: Tested this and can confirm - if the last recommendation step viewed before checking out this PR was the `security-plan` recommendation, an error is thrown indicating "Unknown step" when trying to view the recommendations with the new PR checked out.

**Review by jboland88 — CHANGES_REQUESTED** [review]
Overall, the code looks straightforward, there's a handful of things I think we may need to consider though:

- If the last active step in the recommendations was the old `security-plan` step before this PR is applied, there's an error when trying to view the recommendations again w/o resetting options. I think we can change what path the old slug redirects to, maybe should go to the summary screen?
- Wondering if we should do anything to redirect the old `/recommendations/security-plan` URL. Maybe not, the default is to redirect to the Jetpack dashboard IIRC.
- I think we may need to re-evaluate the conditions that enable the protect conditional recommendation in the `plugin_activated` function in `class-jetpack-recommendations.php` file. If the protect plugin is already installed, probably would not make sense to recommend it :) I think there's some more details here: pbNhbs-3nL-p2

File: projects/plugins/jetpack/_inc/client/recommendations/constants.js
Comment (CodeyGuyDylan) [reply projects/plugins/jetpack/_inc/client/recommendations/constants.js]: I would have thought the point of having constants like this is so that if we have to change the value, we don't also have to go change everywhere where the constant is used. Is there another reason for it that I am unaware of?

File: projects/plugins/jetpack/_inc/client/recommendations/sidebar/product-spotlight/index.jsx
Comment (CodeyGuyDylan) [reply projects/plugins/jetpack/_inc/client/recommendations/sidebar/product-spotlight/index.jsx]: I agree with you completely, not sure why I went `id` here 😅

File: projects/plugins/jetpack/_inc/lib/class-jetpack-recommendations.php
Comment (CodeyGuyDylan) [reply projects/plugins/jetpack/_inc/lib/class-jetpack-recommendations.php]: Same question as before, though if its not for reusability, I can see how it would still be useful to go and see the one source of truth "value" in a constant such as this. I'll go ahead and update them 😄

File: projects/plugins/jetpack/_inc/client/recommendations/constants.js
Comment (CodeyGuyDylan) [reply projects/plugins/jetpack/_inc/client/recommendations/constants.js:12]: You are correct, I'll get that error handled :)

**Comment by CodeyGuyDylan** [comment]
> If the last active step in the recommendations was the old security-plan step before this PR is applied, there's an error when trying to view the recommendations again w/o resetting options. I think we can change what path the old slug redirects to, maybe should go to the summary screen?

I solved this by adding a case for the old `SECURITY-PLAN` slug and we can add all old recommendations there until we refactor the assistant

> Wondering if we should do anything to redirect the old /recommendations/security-plan URL. Maybe not, the default is to redirect to the Jetpack dashboard IIRC.

That is correct, so I think we are good to leave this as is

> I think we may need to re-evaluate the conditions that enable the protect conditional recommendation in the plugin_activated function in class-jetpack-recommendations.php file. If the protect plugin is already installed, probably would not make sense to recommend it :) I think there's some more details here: pbNhbs-3nL-p2

Thank you for pointing this out! It should now not show if the user has a plan that has Scan, Scan, or Protect

**Comment by CodeyGuyDylan** [comment]
To test out the new conditions, I tried activating a plugin to trigger the Protect recommendation under the following circumstances:
* When I had JP Scan active (it did not show up)
* When I had a JP Security plan (it did not show up)
* When I had a JP Complete plan (it did not show up)
* When I already had Jetpack Protect (it did not show up)
* Without any of the above plans or products (it showed up)

Seems to work well for me

To test out the issue with the old `security-plan` slug being leftover if the last recommendation was `security-plan`, I had `trunk` checked out and triggered the `security-plan` recommendation. I viewed it, and then checked out this branch. I was successfully redirected to the summary page

File: projects/plugins/jetpack/_inc/client/recommendations/index.jsx
Comment (jboland88) [inline projects/plugins/jetpack/_inc/client/recommendations/index.jsx]: This makes me wonder if we should just have the default case be to redirect to the summary screen 🤔 Would be a more graceful way to "fail" and allow us to remove the old step slug/ constant completely instead of needing to retain it just for the redirect.

Thanks for iterating here - did not think of this until I saw the updated version.

What do you think, would that work?

File: projects/plugins/jetpack/_inc/lib/class-jetpack-recommendations.php
Comment (jboland88) [inline projects/plugins/jetpack/_inc/lib/class-jetpack-recommendations.php:217]: Might also be nice to add `jetpack-protect.php` to the whitelist array above. I think you'd catch that by checking with `Plugins_Installer::is_plugin_active` further down, but this would be an earlier, cheaper check.

File: projects/plugins/jetpack/_inc/lib/class-jetpack-recommendations.php
Comment (jboland88) [inline projects/plugins/jetpack/_inc/lib/class-jetpack-recommendations.php]: This comment is slightly confusing - the check below appears to just check to see if the protect plugin is active?

File: projects/plugins/jetpack/_inc/client/state/recommendations/reducer.js
Comment (jboland88) [inline projects/plugins/jetpack/_inc/client/state/recommendations/reducer.js]: I think this should be `jetpack-protect/jetpack-protect.php`

File: projects/plugins/jetpack/_inc/client/state/recommendations/reducer.js
Comment (jboland88) [inline projects/plugins/jetpack/_inc/client/state/recommendations/reducer.js]: I think this should be `jetpack-protect/jetpack-protect.php`

**Review by jboland88 — CHANGES_REQUESTED** [review]
Thanks for the updates! I noticed a few more items, I think the only blockers are the plugin paths/ slugs.

> To test out the issue with the old security-plan slug being leftover if the last recommendation was security-plan, I had trunk checked out and triggered the security-plan recommendation. I viewed it, and then checked out this branch. I was successfully redirected to the summary page

I tried this again - I ended up getting sent to the `site-type` step 🤔 I think I need to have viewed other steps in the recommendations. I'll try this again in a subsequent test, looks like it should work based on the code.

I tested on a site with a Jetpack Security plan and did not get the recommendation 👍

File: projects/plugins/jetpack/_inc/client/recommendations/index.jsx
Comment (CodeyGuyDylan) [reply projects/plugins/jetpack/_inc/client/recommendations/index.jsx]: I am not sure why I didn't think of that either haha, I think redirecting to the summary page by default is the best option, an error that prevents the page from loading is a bad user experience no matter how good the page is haha

File: projects/plugins/jetpack/_inc/lib/class-jetpack-recommendations.php
Comment (CodeyGuyDylan) [reply projects/plugins/jetpack/_inc/lib/class-jetpack-recommendations.php]: Nice catch, that comment was from a previous iteration I was trying haha, I will update it!

File: projects/plugins/jetpack/_inc/client/state/recommendations/reducer.js
Comment (CodeyGuyDylan) [reply projects/plugins/jetpack/_inc/client/state/recommendations/reducer.js]: You are correct! Thank you

File: projects/plugins/jetpack/_inc/client/state/recommendations/reducer.js
Comment (CodeyGuyDylan) [reply projects/plugins/jetpack/_inc/client/state/recommendations/reducer.js]: 😄

File: projects/plugins/jetpack/_inc/client/recommendations/constants.js
Comment (jboland88) [inline projects/plugins/jetpack/_inc/client/recommendations/constants.js]: Since we've moved towards dropping `SECURITY_PLAN` elsewhere, can probably remove it here too.

File: projects/plugins/jetpack/_inc/lib/class-jetpack-recommendations.php
Comment (jboland88) [inline projects/plugins/jetpack/_inc/lib/class-jetpack-recommendations.php]: Not a blocker:
This could be one place where keeping a check for `protect/jetpack-protect.php` would be alright. Would also check for the monorepo version of protect being active before enabling the recommendation.

**Review by jboland88 — DISMISSED** [review]
LGTM and tests well!
For future reviewers, noting again (as mentioned in the test plan) that merging this PR will depend on an update to the Protect plugin.

Thanks for the updates :)
Noted a couple of non-blockers.

**Review by jboland88 — DISMISSED** [review]
Re-approving the small code changes made since my last review.
I have not re-run the test plan for these small changes.

**Comment by sergeymitr** [comment]
Hi @CodeyGuyDylan.

> this PR will have to wait until Protect releases again in order to work correctly

Do you have an ETA on when Protect is going to be released?
If we merge now, this PR will get into Jetpack Beta on Tuesday and go live in a bit over a week.
Should we wait until Protect gets released before merging this PR?

**Comment by CodeyGuyDylan** [comment]
> Do you have an ETA on when Protect is going to be released? If we merge now, this PR will get into Jetpack Beta on Tuesday and go live in a bit over a week. Should we wait until Protect gets released before merging this PR?

@sergeymitr Thank you for asking! We have taken steps to ensure that Protect will be released before the JP Plugin so that this will work. It is in progress now, you can read the slack chat here: p1661282327776849-slack-C029WFNV69M

If you do approve this PR, I will wait until this has officially happened to merge just in case it does not actually get done before the code freeze 😄

**Comment by CodeyGuyDylan** [comment]
FWIW, the reason we are pushing to get this out before the release is that we have 2 other Assistant Recommendations that have already been merged and are completely ready to be in production

**Comment by sergeymitr** [comment]
@CodeyGuyDylan please ping me when Protect gets released, and I'll give the PR a proper review asap.

Discussion about the Protect delay: p1661528496430159-slack-C01U2KGS2PQ

**Comment by CodeyGuyDylan** [comment]
Will do!

**Comment by bindlegirl** [comment]
With the new Protect 1.0.4 being released yesterday, this PR is ready to get merged.

**Review by bindlegirl — APPROVED** [review]
LGTM! PR needed rebasing and resolving a couple of conflicts related to changes added in #25446

**Comment by CodeyGuyDylan** [comment]
Thank you for taking care of this @bindlegirl!

**Pull request merged**
